cmd/stackdriver-prometheus-sidecar/statusz.go

```diff
@@ -0,0 +1,197 @@
+// Copyright 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"fmt"
+	"html/template"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/scrape"
+)
+
+var (
+	serverStart = time.Now()
+)
+
+type statuszHandler struct {
+	logger    log.Logger
+	projectId string
+	cfg       *mainConfig
+}
+
+func (h *statuszHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	var data struct {
+		ServerName    string
+		VersionInfo   string
+		BuildContext  string
+		Uname         string
+		FdLimits      string
+		StartTime     string
+		Uptime        string
+		PodName       string
+		NodeName      string
+		NamespaceName string
+		GKEInfo       struct {
+			ProjectId       string
+			ClusterLocation string
+			ClusterName     string
+		}
+		DisplayConfig  map[string]string
+		MetricRenames  map[string]string
+		StaticMetadata []scrape.MetricMetadata
+	}
+
+	data.ServerName = filepath.Base(os.Args[0])
+
+	data.VersionInfo = version.Info()
+	data.BuildContext = version.BuildContext()
+	data.Uname = Uname()
+	data.FdLimits = FdLimits()
+
+	data.StartTime = serverStart.Format(time.RFC1123)
+	uptime := int64(time.Since(serverStart).Seconds())
+	data.Uptime = fmt.Sprintf("%d hr %02d min %02d sec",
+		uptime/3600, (uptime/60)%60, uptime%60)
+
+	data.PodName = os.Getenv("POD_NAME")
+	data.NodeName = os.Getenv("NODE_NAME")
+	data.NamespaceName = os.Getenv("NAMESPACE_NAME")
+
+	data.GKEInfo.ProjectId = h.projectId
+	data.GKEInfo.ClusterLocation = h.cfg.kubernetesLabels.location
+	data.GKEInfo.ClusterName = h.cfg.kubernetesLabels.clusterName
+
+	data.DisplayConfig = map[string]string{
```

Comment:
Please Implement String() for type of `h.cfg` , otherwise the code here will get out-of-sync easily. The corresponding template should just leverage the String() method of `h.cfg`

Reply:
Moved to a function named TableForStatusz().